pkg/reconciler/tenancy/workspacemounts/workspace_indexes.go

```diff
@@ -51,33 +56,72 @@ type workspaceMountsReferenceKey struct {
 	Namespace   string `json:"namespace,omitempty"`
 }
 
-func indexWorkspaceByMountObject(obj interface{}) ([]string, error) {
-	ws, ok := obj.(*tenancyv1alpha1.Workspace)
-	if !ok {
-		return []string{}, fmt.Errorf("obj is supposed to be a Workspace, but is %T", obj)
-	}
-
-	if ws.Spec.Mount == nil {
-		return nil, nil
+func retryRESTMapping(forCluster *dynamicrestmapper.ForCluster, gvk schema.GroupVersionKind) (schema.GroupVersionResource, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			return schema.GroupVersionResource{}, fmt.Errorf("timeout while retrying REST mapping for %s", gvk)
+		default:
+			mapping, err := forCluster.RESTMapping(gvk.GroupKind(), gvk.Version)
+			if err == nil {
+				return mapping.Resource, nil
+			}
+			if meta.IsNoMatchError(err) {
+				// If the error is a NoMatchError, we can retry the mapping
+				// after a short delay.
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+		}
 	}
+}
 
-	key := workspaceMountsReferenceKey{
-		ClusterName: logicalcluster.From(ws).String(),
-		// TODO(sttts): do proper REST mapping
-		Resource:  strings.ToLower(ws.Spec.Mount.Reference.Kind) + "s",
-		Name:      ws.Spec.Mount.Reference.Name,
-		Namespace: ws.Spec.Mount.Reference.Namespace,
-	}
-	cs := strings.SplitN(ws.Spec.Mount.Reference.APIVersion, "/", 2)
-	if len(cs) == 2 {
-		key.Group = cs[0]
+func newIndexWorkspaceByMountObject(dynamicRESTMapper *dynamicrestmapper.DynamicRESTMapper) cache.IndexFunc {
+	return func(obj interface{}) ([]string, error) {
+		ws, ok := obj.(*tenancyv1alpha1.Workspace)
+		if !ok {
+			return []string{}, fmt.Errorf("obj is supposed to be a Workspace, but is %T", obj)
+		}
+
+		if ws.Spec.Mount == nil {
+			return nil, nil
+		}
+
+		gv, err := schema.ParseGroupVersion(ws.Spec.Mount.Reference.APIVersion)
+		if err != nil {
+			return nil, err // TODO flesh out error
+		}
+		gvk := schema.GroupVersionKind{
+			Group:   gv.Group,
+			Version: gv.Version,
+			Kind:    ws.Spec.Mount.Reference.Kind,
+		}
+
+		forCluster := dynamicRESTMapper.ForCluster(logicalcluster.From(ws))
+		gvr, err := retryRESTMapping(forCluster, gvk)
+		if err != nil {
+			return nil, err // TODO flesh out error
```

Comment:
```suggestion
			return nil, err
```

Since retryRESTMapping is private I think it doesn't need an additional error.